src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationAuthHandler.cs

```diff
@@ -0,0 +1,104 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Encodings.Web;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters.Authentication;
+
+/// <summary>
+/// Authentication handler that authenticates users by making requests to a remote app
+/// for authentication via a remote authentication service.
+/// </summary>
+internal partial class RemoteAppAuthenticationAuthHandler : AuthenticationHandler<RemoteAppAuthenticationOptions>
+{
+    private readonly IRemoteAppAuthenticationService _authService;
+    private readonly IEnumerable<IRemoteAppAuthenticationResultProcessor> _resultProcessors;
+    private readonly ILogger<RemoteAppAuthenticationAuthHandler> _logger;
+    private RemoteAppAuthenticationResult? _remoteAppAuthResult;
+
+    public RemoteAppAuthenticationAuthHandler(IRemoteAppAuthenticationService authService,
+                                           IEnumerable<IRemoteAppAuthenticationResultProcessor> resultProcessors,
+                                           IOptionsMonitor<RemoteAppAuthenticationOptions> options,
+                                           ILoggerFactory loggerFactory,
+                                           UrlEncoder encoder,
+                                           ISystemClock clock)
+        : base(options, loggerFactory, encoder, clock)
+    {
+        _logger = loggerFactory.CreateLogger<RemoteAppAuthenticationAuthHandler>();
+        _authService = authService;
+        _resultProcessors = resultProcessors;
+    }
+
+    protected override Task InitializeHandlerAsync() => _authService.InitializeAsync(Scheme);
+
+    private async Task<RemoteAppAuthenticationResult> GetRemoteAppAuthenticationResultAsync()
+    {
+        if (_remoteAppAuthResult is null)
```

Comment:
We cache the result because we need the remote app authentication result both for `HandleAuthenticateAsync` and for `HandleChallengeAsync`, so this may be called twice.